net/server.go

```diff
@@ -196,7 +195,7 @@ func (s *server) addPubSubTopic(topic string, subscribe bool) error {
 		}
 	}
 
-	t, err := rpc.NewTopic(s.peer.ctx, s.peer.ps, s.peer.host.ID(), topic, subscribe)
+	t, err := rpc.NewTopic(context.Background(), s.peer.ps, s.peer.host.ID(), topic, subscribe)
```

Comment:
question: Are you sure this is an improvement?  It does look like a suspicious change to me, I would have expected us to want to use the same context for all topics+peer

Reply:
I've always followed this rule from the Go docs:

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

Since there is no context to pass around in these functions they should be run with a background context..

Reply:
In that case I really think we should ignore that suggestion here and use the context provided to the system.    Please consider reverting this change a todo.

Reply:
I don't think the context provided to the constructor makes sense either since it would outlive the function call. Would you be okay with creating a new background context that is then used for all publish / topic calls?

Reply:
> I don't think the context provided to the constructor makes sense either since it would outlive the function call.

Why does that not make sense?  Any context provided to a function outlives the function call

Reply:
This example shows how the shared context is confusing for callers:

```go
func setupNetworkWithTimeout(ctx context.Context) (*net.Peer, error) {
  // cancel setup if it takes too long
  ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
  defer cancel()

  // context is shared in peer struct with all sub functions
  n, err := net.NewPeer(ctx, ...)
  if err != nil {
    return nil, err
  }

  return n, nil
}

func main() {
  ctx := context.Background()

  n, err := setupNetworkWithTimeout(ctx)
  if err != nil {
    panic(err)
  }
  
  // the shared context was cancelled so the peer cannot run
  n.Start()
}
```

Reply:
I mean if I cancelled a context that I used to construct something, I'd want it to close all the things owned by that.  I like that about the current setup.

It also provides all the other stuff that context holds to the children of `Peer`, if nothing else it is handy for logging.

Reply:
suggestion: If you are not happy about this, I suggest we discuss it in the standup.

Reply:
> I mean if I cancelled a context that I used to construct something, I'd want it to close all the things owned by that. I like that about the current setup.

If we didn't have explicit `Start` and `Stop` functions I would agree with this.

Reply:
> If we didn't have explicit Start and Stop functions I would agree with this.

I could rant for quite a while about how much I hate `Start` functions on objects with constructors...  I don't see a `Stop` function though, although I could rant about that/those paired with `Close` functions too :sweat_smile:.

Reply:
Interesting discussion, I feel like we have discussed something similar before. While I do personally have a slight preference for Andy's approach. Since I read the Go guidelines regarding not storing contexts in structs I do tend to stick to those guidelines and am happy with this change. However I do wonder if it's possible to just pass the ctx to `addPubSubTopic` through a function parameter (most of the callers do have a `ctx`).

Reply:
> However I do wonder if it's possible to just pass the ctx to addPubSubTopic through a function parameter (most of the callers do have a ctx).

I'm open to this as long as that context is not the same as the constructor context.